les/api_backend.go

```diff
@@ -17,128 +17,202 @@
+
+func (eth *LightEthereum) HeaderByHash(ctx context.Context, blockhash common.Hash) (*types.Header, error) {
+	if h := eth.blockchain.GetHeaderByHash(blockhash); h != nil {
+		return h, nil
+	}
+	return nil, ErrBlockNotFound
 }
 
-func (b *LesApiBackend) SetHead(number uint64) {
-	b.eth.blockchain.SetHead(number)
+func (eth *LightEthereum) HeaderByNumber(ctx context.Context, blocknum *big.Int) (*types.Header, error) {
+	if blocknum == nil {
+		return eth.blockchain.CurrentHeader(), nil
+	}
+	h, err := eth.blockchain.GetHeaderByNumberOdr(ctx, uint64(blocknum.Uint64()))
+	if err != nil {
+		return nil, err
+	}
+	if h == nil {
+		return nil, ErrBlockNotFound
+	}
+	return h, nil
 }
 
-func (b *LesApiBackend) HeaderByNumber(ctx context.Context, blockNr rpc.BlockNumber) (*types.Header, error) {
-	if blockNr == rpc.LatestBlockNumber || blockNr == rpc.PendingBlockNumber {
-		return b.eth.blockchain.CurrentHeader(), nil
+func (eth *LightEthereum) BlockByNumber(ctx context.Context, blocknum *big.Int) (*types.Block, error) {
+	header, err := eth.HeaderByNumber(ctx, blocknum)
+	if err != nil {
+		return nil, err
 	}
+	return eth.BlockByHash(ctx, header.Hash())
+}
 
-	return b.eth.blockchain.GetHeaderByNumberOdr(ctx, uint64(blockNr))
+func (eth *LightEthereum) BlockByHash(ctx context.Context, blockhash common.Hash) (*types.Block, error) {
+	return eth.blockchain.GetBlockByHash(ctx, blockhash)
 }
 
-func (b *LesApiBackend) BlockByNumber(ctx context.Context, blockNr rpc.BlockNumber) (*types.Block, error) {
-	header, err := b.HeaderByNumber(ctx, blockNr)
-	if header == nil || err != nil {
+func (eth *LightEthereum) BlockReceipts(ctx context.Context, blockhash common.Hash, blocknum uint64) ([]*types.Receipt, error) {
+	return light.GetBlockReceipts(ctx, eth.odr, blockhash, blocknum)
+}
+
+func (eth *LightEthereum) BlockTD(blockHash common.Hash) *big.Int {
+	return eth.blockchain.GetTdByHash(blockHash)
+}
+
+func (eth *LightEthereum) TransactionByHash(ctx context.Context, txHash common.Hash) (tx *types.Transaction, isPending bool, err error) {
+	if tx = eth.txPool.GetTransaction(txHash); tx != nil {
+		return tx, true, nil
+	}
+	return nil, false, ErrTxNotFound
+}
+
+func (eth *LightEthereum) TransactionInBlock(ctx context.Context, blockhash common.Hash, index uint) (*types.Transaction, error) {
+	b, err := eth.blockchain.GetBlockByHash(ctx, blockhash)
+	if err != nil {
 		return nil, err
 	}
-	return b.GetBlock(ctx, header.Hash())
+	if index >= uint(len(b.Transactions())) {
+		return nil, fmt.Errorf("transaction index out of range")
+	}
+	return b.Transactions()[index], nil
 }
 
-func (b *LesApiBackend) StateAndHeaderByNumber(ctx context.Context, blockNr rpc.BlockNumber) (ethapi.State, *types.Header, error) {
-	header, err := b.HeaderByNumber(ctx, blockNr)
-	if header == nil || err != nil {
-		return nil, nil, err
+func (eth *LightEthereum) TransactionReceipt(ctx context.Context, txhash common.Hash) (*types.Receipt, error) {
+	return nil, ErrReceiptNotFound
+}
```

Comment:
This is very unfortunate. Is there any way I can make this method work with the current les protocol?